### PR TITLE
Ensure @AfterGroups(alwaysRun=true) always runs

### LINKED ISSUE #1880: @AfterGroups(alwaysRun = true) is not called if there is an exception in @BeforeGroups
### TestNG Version
As far as I know, this issue has been existed since 6.9.10.

### Expected behavior
@AfterGroups(alwaysRun = true) is called even if test is throwing exception in @BeforeGroups.

### Actual behavior
@AfterGroups(alwaysRun = true) is not called

### Is the issue reproductible on runner?

- [ ] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [X] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample

```java
public class TestAfterGroups {
    @BeforeGroups(groups = "123")
    public void before() throws Exception {
        throw new Exception("before groups fail");
    }

    @org.testng.annotations.Test(groups = "123")
    public void test() {

    }

    @AfterGroups(groups = "123", alwaysRun = true)
    public void after() {
        System.out.println("after groups");//never invoked in case of exception
    }

}
```

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1880

Fixes #1880 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/internal/annotations/AnnotationHelper.java [hunk 1]
```diff
@@ -163,10 +163,12 @@ private static IConfigurationAnnotation createConfiguration(
     }
     if (bg != null) {
       result.setBeforeGroups(bg.getBeforeGroups());
+      result.setBeforeGroups(true);
```
Comment (juherr) [inline src/main/java/org/testng/internal/annotations/AnnotationHelper.java]: Code smells: 2 methods with the same name but different signature

File: src/main/java/org/testng/internal/annotations/ConfigurationAnnotation.java [hunk 2]
```diff
@@ -136,6 +138,24 @@ public void setAfterGroups(String[] afterGroups) {
     m_afterGroups = afterGroups;
```
Comment (juherr) [inline src/main/java/org/testng/internal/annotations/ConfigurationAnnotation.java]: Try to add here: `this.afterGroups = true`

File: src/main/java/org/testng/internal/annotations/ConfigurationAnnotation.java [hunk 3]
```diff
@@ -27,6 +27,8 @@
   private boolean m_beforeSuite = false;
   private boolean m_afterSuite = false;
   private boolean m_alwaysRun = false;
+  private boolean beforeGroups = false;
```
Comment (juherr) [inline src/main/java/org/testng/internal/annotations/ConfigurationAnnotation.java]: Doc is missing: what is the difference between `beforeGroups` and `m_beforeGroups`?

File: src/main/java/org/testng/annotations/IConfigurationAnnotation.java [hunk 4]
```diff
@@ -58,4 +58,12 @@
    *     a @BeforeSuite or similar that is represented as a configuration annotation.
    */
   boolean isFakeConfiguration();
+
+  default boolean isBeforeGroups() {
```
Comment (juherr) [inline src/main/java/org/testng/annotations/IConfigurationAnnotation.java]: Not clear what is mean.

**Review by juherr — COMMENTED** [review]
The fix is working but I'm not sure to understand it. Some explanations are missing.

File: src/main/java/org/testng/internal/annotations/AnnotationHelper.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/annotations/AnnotationHelper.java]: @juherr - I didn't have a choice. The `setBeforeGroups(boolean)` is basically being used to flag that an annotation is a `@BeforeGroups` representation. The name is already taken. For lack of a better nomenclature, I just decided to overload the method and provide another variant.

**Comment by krmahadevan** [comment]
@juherr - I took another closer look at the current implementation. Turns out that I didn't need the new booleans after all. I have amended the commit and pushed in the changes. Please help take a look.

**Review by juherr — APPROVED** [review]

**Comment by krmahadevan** [comment]
Based on @cbeust email, I am going ahead with merging this since @juherr has approved the changes.

**Pull request merged**